src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,112 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int INITIAL_SIZE = 10;
+    private static final int EMPTY_SIZE = 0;
+    private static final int CAPACITY_DIVISOR = 2;
+    private T[] dataElement;
+    private int size;
+
+    public ArrayList() {
+        dataElement = (T[]) new Object[INITIAL_SIZE];
+    }
+
     @Override
     public void add(T value) {
-
+        if (checkCapacity()) {
+            grow();
+        }
+        dataElement[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Can't add to index " + index);
+        }
+        if (checkCapacity()) {
+            grow();
+        }
+        System.arraycopy(dataElement, index, dataElement, index + 1, size++ - index);
+        dataElement[index] = value;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkArrayListIndexOutOfBoundsException(index, "get");
+        return dataElement[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkArrayListIndexOutOfBoundsException(index, "set");
+        dataElement[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkArrayListIndexOutOfBoundsException(index, "remove");
+        final T element = dataElement[index];
+        System.arraycopy(dataElement, index + 1, dataElement, index, size-- - index - 1);
+        return element;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (element == dataElement[i] || areEqual(element,dataElement[i])) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Not found element in the list");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == EMPTY_SIZE;
+    }
+
+    private void grow() {
+        int oldCapacity = dataElement.length;
+        int newCapacity = oldCapacity + (oldCapacity / CAPACITY_DIVISOR);
+        if (newCapacity < 0) {
+            newCapacity = Integer.MAX_VALUE;
+        }
+        T[] newDataElement = (T[]) new Object[newCapacity];
+        System.arraycopy(dataElement, 0, newDataElement, 0, oldCapacity);
+        dataElement = newDataElement;
+    }
+
+    private boolean checkIndex(int index) {
+        return index < 0 || index >= size;
+    }
+
+    private boolean areEqual(T a, T b) {
+        return (a != null && b != null && a.hashCode() == b.hashCode());
```

Comment:
what about collision? when non-equal objects have the same hashcode?

Reply:
overrided hashcode and made prime number generation for more collision safety